File: JavaScript/linked_list/linked_list.js
class Node {
    // Constructors
    constructor() {
        this._data = 0;
        this._next = null;
    }

    constructor(data) {
        this._data = data;
        this._next = null;
    }

    constructor(data, next) {
        this._data = data;
        this._next = next;
    }
}

class LinkedList {
    // Constructors
    constructor() {
        this._head = null;
        this._size = 0;
    }

    // Public Methods
    append(data) {
        if (this._head == null) {
            this._head = Node(data);
            this._size++;
            return;
        }

        let current = this._head;
        while (current._next != null) {
            current = current._next;
        }

        current._next = Node(data);
        this._size++;
    }

    prepend(data) {
        if (this._head == null) {
            this._head = Node(data);
            this._size++;
            return;
        }

        const newHead = Node(data);
        newHead._next = this._head;
        this._head = newHead;
        this._size++;
    }

    insert(data, index) {
        if (index < 1) {
            this.prepend(data);
            return;
        }
        else if (index >= this._size) {
            this.append(data);
            return;
        }

        let current = this._head;
        for (let i = 0; i < index - 1; i++) {
            current = current._next;
        }

        let newNode = Node(data);
        newNode._next = current._next;
        current._next = newNode;
        this._size++;
    }

    remove(data) {
        if (this._head == null) {
            return;
        }

        if (this._head._data == data) {
            this._head = this._head._next;
            this._size--;
            return;
        }

        let current = this._head;
        while (current._next != null) {
            if (current._next._data == data) {
                current._next = current._next._next;
                this._size--;
                return;
            }
            current = current._next;
        }
    }

    contains(data) {
        let current = this._head;
        while (current != null) {
            if (current._data == data) {
                return true;
            }
            current = current._next;
        }
    }

    print() {
        let current = this._head;
        while (current != null) {
            console.log(current._data);
            current = current._next;
        }
    }
}

File: Python/Linked_List/LinkedList.py
class Node:
    # Constructors
    def __init__(self):
        self._data = 0
        self._next = None

    def __init__(self, data):
        self._data = data
        self._next = None

    def __init__(self, data, next):
        self._data = data
        self._next = next


class LinkedList:
    # Contstructors
    def __init__(self):
        self._head = None
        self._size = 0

    def __init__(self, head):
        self._head = head
        self._size = 0

    def __init__(self, list):
        self._head = Node(list[0])
        self._size = 1

        current = self._head
        for i in range(1, len(list)):
            current._next = Node(list[i])
            current = current.getNext()
            self._size += 1

    # Methods
    def push_front(self, data):
        self._head = Node(data, self._head)

    def push_back(self, data):
        current = self._head

        while current._next() != None:
            current = current._next

        current._next = Node(data)

    def insert(self, data, idx):
        if idx <= 0:
            self.push_front(data)
        elif idx >= self._size:
            self.push_back(data)
        else:
            current = self._head
            for i in range(idx):
                current = current._next

            current._next = Node(data, current._next)

    def remove(self, data):
        current = self._head
        while current._next != None:
            if current._next.data == data:
                current._next = current._next.next
                return
            current = current._next

    def contains(self, data):
        current = self._head
        while current._next != None:
            if current._data == data:
                return True
            current = current._next
        return False

    def print(self):
        current = self._head
        while current != None:
            print(current._data + " ")
            current = current._next
